pkg/auth/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -0,0 +1,84 @@
+package keystonepassword
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"runtime/debug"
+
+	"github.com/golang/glog"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
+	"github.com/rackspace/gophercloud/openstack/identity/v3/tokens"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+	"github.com/openshift/origin/pkg/auth/authenticator"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+// keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
+type keystonePasswordAuthenticator struct {
+	providerName   string
+	url            string
+	client         *http.Client
+	domainName     string
+	identityMapper authapi.UserIdentityMapper
+}
+
+// New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
+// A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
+// If no transport is provided, http.DefaultTransport is used
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+	if transport == nil {
+		transport = http.DefaultTransport
+	}
+	client := &http.Client{Transport: transport}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+}
+
+// AuthenticatePassword approves any login attempt which is successfully validated with Keystone
+func (a keystonePasswordAuthenticator) AuthenticatePassword(username, password string) (user.Info, bool, error) {
+	defer func() {
+		if e := recover(); e != nil {
+			util.HandleError(fmt.Errorf("Recovered panic: %v, %s", e, debug.Stack()))
+		}
+	}()
+
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: a.url,
+		Username:         username,
+		Password:         password,
+		DomainName:       a.domainName,
+	}
+
+	// Calling NewClient/Authenticate manually rather than simply calling AuthenticatedClient
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
+	client.HTTPClient = *a.client
+	if err == nil {
+		err = openstack.AuthenticateV3(client, opts)
+	}
+
+	if err != nil {
+		glog.Infof("Failed: Starting openstack authenticate client: %v", err)
+		if responseErr, ok := err.(*gophercloud.UnexpectedResponseCodeError); ok {
+			if responseErr.Actual == 401 {
+				return nil, false, nil
+			}
+		}
+		if err == tokens.ErrMissingPassword {
```

Comment:
It's not valid. The gophercloud library errors out with ErrMissingPassword without even making the call to keystone, but yeah if we do the same check ourselves it might make more sense than passing it on and checking for the error return.